File: Algorithms/Sorting/Lily's Homework.py
import os

#
# Complete the 'lilysHomework' function below.
#
# The function is expected to return an INTEGER.
# The function accepts INTEGER_ARRAY arr as parameter.
#

def lilysHomework(arr):
    # Write your code here
    def count_swaps(sorted_arr):
        swaps = 0
        visited = {i: False for i in range(len(arr))}
        value_to_index = {v: i for i, v in enumerate(sorted_arr)}

        for i, v in enumerate(sorted_arr):
            if visited[i] or v == arr[i]:
                continue

            j = i
            cycle_size = 0
            while not visited[j]:
                visited[j] = True
                j = value_to_index[arr[j]]
                cycle_size += 1

            if cycle_size > 1:
                swaps += cycle_size - 1

        return swaps

    ascending_arr = sorted(arr)
    descending_arr = sorted(arr, reverse=True)

    return min(count_swaps(ascending_arr), count_swaps(descending_arr))

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input().strip())

    arr = list(map(int, input().rstrip().split()))

    result = lilysHomework(arr)

    fptr.write(str(result) + '\n')
    fptr.close()

File: Algorithms/Sorting/Lily's Homework.php
<?php

/*
 * Complete the 'lilysHomework' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts INTEGER_ARRAY arr as parameter.
 */

function lilysHomework($arr) {
    # Write your code here
    function countSwaps($arr, $sortedArr) {
        $swaps = 0;
        $n = count($arr);
        $visited = array_fill(0, $n, False);
        $valueToIndex = array_flip($sortedArr);

        for ($i = 0; $i < $n; $i++) {
            if ($visited[$i] || $arr[$i] == $sortedArr[$i]) {
                continue;
            }

            $cycleSize = 0;
            $j = $i;

            while (!$visited[$j]) {
                $visited[$j] = True;
                $j = $valueToIndex[$arr[$j]];
                $cycleSize++;
            }

            if ($cycleSize > 1) {
                $swaps += $cycleSize - 1;
            }
        }

        return $swaps;
    }

    $ascendingArr = $arr;
    sort($ascendingArr);

    $descendingArr = $arr;
    rsort($descendingArr);

    return min(countSwaps($arr, $ascendingArr), countSwaps($arr, $descendingArr));
}


$fptr = fopen(getenv("OUTPUT_PATH"), "w");

$n = intval(trim(fgets(STDIN)));
$arr_temp = rtrim(fgets(STDIN));
$arr = array_map('intval', preg_split('/ /', $arr_temp, -1, PREG_SPLIT_NO_EMPTY));

$result = lilysHomework($arr);

fwrite($fptr, $result . "\n");
fclose($fptr);
